native/0004-envoy-url-socks5-param.patch

```diff
@@ -0,0 +1,101 @@
++    }
++  }
+   // Initialize context on the init thread.
+-  if (params->envoy_url.rfind("socks5://", 0) != 0) {
+-  cronet::PostTaskToInitThread(
++  if (!envoy_socks.is_empty()) {
++    cronet::PostTaskToInitThread(
++      FROM_HERE,
++      base::BindOnce(&CronetContext::InitRequestContextOnInitThreadWithUri,
++                     base::Unretained(context_.get()), envoy_socks.spec()));
++  } else if (params->envoy_url.rfind("socks5://", 0) == 0) {
++    // socks5:// URL
++    cronet::PostTaskToInitThread(
++      FROM_HERE,
++      base::BindOnce(&CronetContext::InitRequestContextOnInitThreadWithUri,
++                     base::Unretained(context_.get()), params->envoy_url));
++  } else {
++    cronet::PostTaskToInitThread(
+       FROM_HERE, base::BindOnce(&CronetContext::InitRequestContextOnInitThread,
+                                 base::Unretained(context_.get())));
+-  } else {
+-      cronet::PostTaskToInitThread(
+-          FROM_HERE,
+-          base::BindOnce(&CronetContext::InitRequestContextOnInitThreadWithUri,
+-                         base::Unretained(context_.get()), params->envoy_url));
+   }
+   return CheckResult(Cronet_RESULT_SUCCESS);
+ }
+diff --git a/components/cronet/native/sample/main.cc b/components/cronet/native/sample/main.cc
+index ffdb1a37fd629..007d5bbab90d7 100644
+--- a/components/cronet/native/sample/main.cc
++++ b/components/cronet/native/sample/main.cc
+@@ -12,6 +12,7 @@ Cronet_EnginePtr CreateCronetEngine() {
+   Cronet_EnginePtr cronet_engine = Cronet_Engine_Create();
+   Cronet_EngineParamsPtr engine_params = Cronet_EngineParams_Create();
+   Cronet_EngineParams_user_agent_set(engine_params, "CronetSample/1");
++
+   Cronet_EngineParams_envoy_url_set(engine_params,
+                                     "https://example.com/enovy_path/");
+   Cronet_EngineParams_envoy_url_set(
+@@ -36,7 +37,13 @@ Cronet_EnginePtr CreateCronetEngine() {
+       "?url=https%3A%2F%2Fexample.com%2Fenvoy_path%2F%3Fk1%3Dv1&header_Host="
+       "subdomain.example.com&address=1.2.3.4&disabled_cipher_suites=0xc024,0xc02f");
+   Cronet_EngineParams_envoy_url_set(engine_params, "socks5://127.0.0.1:1080");
+-
++  // proxy URL and SOCKS5 together (for true PTs)
++  Cronet_EngineParams_envoy_url_set(
++      engine_params,
++      "envoy://"
++      "?url=https%3A%2F%2Frayon.example.com%2Fwikipedia%2F&address=142.65.13.41"
++      "&header_Host=abc.example.com&socks5=socks5%3A%2F%2Flocalhost%3A8192");
++ 
+   Cronet_EngineParams_enable_quic_set(engine_params, true);
+ 
+   Cronet_Engine_StartWithParams(cronet_engine, engine_params);
+@@ -73,7 +80,7 @@ int main(int argc, const char* argv[]) {
+   std::cout << "Cronet version: "
+             << Cronet_Engine_GetVersionString(cronet_engine) << std::endl;
+ 
+-  std::string url(argc > 1 ? argv[1] : "https://www.example.com");
++  std::string url(argc > 1 ? argv[1] : "https://www.google.com/generate_204");
```

Comment:
did you want to keep this change or was this changed because it's the url we test with?

Reply:
yes and yes. Is there some reason I shouldn't change it?

Reply:
that's fine, i just wondered if it was meant as a temporary change to match our code.

Reply:
No, I did it on purpose. The `cronet_sample` binary is really useful for testing Envoy changes, and the less I have to edit it, the better ;-) And it's a better example with a real world URL